pkg/csi/service/vanilla/controller.go

```diff
@@ -1137,15 +1139,121 @@ func (c *controller) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshot
+	}
+
+	// the returned snapshotID below is a combination of CNS VolumeID and CNS SnapshotID concatenated by the "+" sign.
+	// That is, a string of "<UUID>+<UUID>"
+	snapshotID, snapshotCreateTime, err := common.CreateSnapshotUtil(ctx, c.manager, volumeID, req.Name)
+	if err != nil {
+		return nil, logger.LogNewErrorCodef(log, codes.Internal,
+			"failed to create snapshot on volume %q: %v", volumeID, err)
+	}
+
+	snapshotCreateTimeInProto := timestamppb.New(*snapshotCreateTime)
+	if err := snapshotCreateTimeInProto.CheckValid(); err != nil {
+		return nil, logger.LogNewErrorCodef(log, codes.Internal,
+			"failed to convert creation timestamp to proto format: %v", err)
+	}
+
+	createSnapshotResponse := &csi.CreateSnapshotResponse{
+		Snapshot: &csi.Snapshot{
+			SizeBytes:      snapshotSize * common.MbInBytes,
+			SnapshotId:     snapshotID,
+			SourceVolumeId: volumeID,
+			CreationTime:   snapshotCreateTimeInProto,
+			ReadyToUse:     true,
+		},
+	}
+
+	log.Infof("CreateSnapshot succeeded for snapshot %s "+
+		"on volume %s size %d Time proto %d Timestamp %+v Response: %+v",
+		snapshotID, volumeID, snapshotSize*common.MbInBytes, snapshotCreateTimeInProto,
+		*snapshotCreateTime, createSnapshotResponse)
+	return createSnapshotResponse, nil
 }
 
 func (c *controller) DeleteSnapshot(ctx context.Context, req *csi.DeleteSnapshotRequest) (
 	*csi.DeleteSnapshotResponse, error) {
 	ctx = logger.NewContextWithLogger(ctx)
 	log := logger.GetLogger(ctx)
 	log.Infof("DeleteSnapshot: called with args %+v", *req)
-	return nil, logger.LogNewErrorCode(log, codes.Unimplemented, "deleteSnapshot")
+
+	isBlockVolumeSnapshotEnabled := commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.BlockVolumeSnapshot)
+	if !isBlockVolumeSnapshotEnabled {
+		return nil, logger.LogNewErrorCode(log, codes.Unimplemented, "deleteSnapshot")
+	}
+
+	// Validate arguments
+	// The expected format of the SnapshotId in the DeleteSnapshotRequest is,
+	// a combination of CNS VolumeID and CNS SnapshotID concatenated by the "+" sign.
+	// That is, a string of "<UUID>+<UUID>"
+	csiSnapshotID := req.GetSnapshotId()
+	if len(csiSnapshotID) == 0 {
+		return nil, logger.LogNewErrorCode(log, codes.InvalidArgument,
+			"DeleteSnapshot Snapshot ID must be provided")
+	}
+
+	// Decompose csiSnapshotID based on the expected format mentioned above
+	IDs := strings.Split(csiSnapshotID, common.VSphereCSISnapshotIdDelimiter)
+	if len(IDs) == 0 || len(IDs) != 2 {
+		return nil, logger.LogNewErrorCode(log, codes.InvalidArgument,
+			"DeleteSnapshot provided invalid Snapshot ID")
+	}
+
+	volumeID := IDs[0]
+	snapshotID := IDs[1]
+	log.Debugf("DeleteSnapshot: Deleting snapshot %q on volume %q", snapshotID, volumeID)
+
+	err := common.DeleteSnapshotUtil(ctx, c.manager, volumeID, snapshotID)
+	if err != nil {
+		return nil, logger.LogNewErrorCodef(log, codes.Internal, "Failed to delete snapshot %q on volume %q. Error: %+v", snapshotID, volumeID, err)
+	}
+
```

Comment:
Is there a plan to handle metrics as well?

Reply:
You could push that in this change.

Reply:
Done